torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -920,3 +947,378 @@ def get_rule(
+        type_promotion_table: Type promotion table.
+
+    Returns:
+        Type promotion rule for the node.
+    """
+    op = node.target
+    if not isinstance(op, torch._ops.OpOverload):
+        # TODO(bowbao): diagnostic.emit api.
+        diagnostic.with_additional_message(
+            f"Skipped. node.target is not OpOverload. Got type: {type(op)}"
+        )
+        return None
+    if (rule := type_promotion_table.get_rule(op.overloadpacket)) is None:
+        diagnostic.with_additional_message(
+            f"Skipped. Cannot find type promotion rule for op: {op}"
+        )
+        return None
+
+    diagnostic.with_additional_message(f"Found type promotion rule: {rule}")
+    return rule
+
+
+class _OpTraceDispatchMode(_python_dispatch.TorchDispatchMode):
+    """Trace ops that were dispatched."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.traced_ops = []
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        self.traced_ops.append(func)
+        return func(*args, **kwargs)
+
+
+@_beartype.beartype
+def find_compatible_op_overload(
+    node: torch.fx.Node, args: tuple, kwargs: dict
+) -> torch._ops.OpOverload:
+    """Find the compatible op overload for a node with a given set of args and kwargs.
+
+    Returns:
+        Compatible op overload for the node.
+
+    Raises:
+        RuntimeError: If no compatible op overload is found.
+        AssertionError: If compatible op overload is not an instance of torch._ops.OpOverload.
+        AssertionError: If op is being decomposed.
+    """
+    target = node.target
+    assert isinstance(
+        target, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(target)}"
+    op = target.overloadpacket
+    op_trace_dispatch_mode = _OpTraceDispatchMode()
+    with op_trace_dispatch_mode:
+        op(*args, **kwargs)
+    assert (
+        len(op_trace_dispatch_mode.traced_ops) >= 1
+    ), "Expected at least 1 traced op, got 0"
+
+    new_op_overload = op_trace_dispatch_mode.traced_ops[0]
+    assert isinstance(
+        new_op_overload, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(new_op_overload)}"
+    assert (
+        new_op_overload.overloadpacket == op
+    ), f"Expected same OpOverload packet, got {new_op_overload.overloadpacket} != {op}"
+
+    return new_op_overload
+
+
+@dataclasses.dataclass
+class NodeMeta:
+    """Required meta data for fx.Node during onnx export.
+
+    For each field, it directly maps to `fx.Node.meta` via `meta[field.name] = field.value`.
+    """
+
+    stack_trace: str
+    nn_module_stack: Any
+
+    def assign_to_node(self, node: torch.fx.Node) -> None:
+        """Assigns the meta data to the given node by updating `node.meta`."""
+        for field in dataclasses.fields(self):
+            setattr(node.meta, field.name, getattr(self, field.name))
+
+
+class ExplicitTypePromotionPass(_pass.Transform, torch.fx.Interpreter):
```

Comment:
wow, multiple inheritance. We certainly need a lot of docstring here to describe what and how things are being done. examples appreciated :)

Reply:
I will separate it. Thinking again, it is better at least in this PR to keep them separated.